zivid_description/urdf/macros/zivid_camera.xacro

```diff
@@ -32,13 +32,13 @@
 
         <!-- Zivid Optical (Measurement) and Projector Joints -->
         <joint name="${prefix}optical_joint" type="fixed">
-            <origin xyz="0.065 0.062 0.0445" rpy="-${0.5*M_PI} 0 -${0.5*M_PI + 8.5/180*M_PI}"/>
+            <origin xyz="0.065 0.062 0.0445" rpy="-${0.5*pi} 0 -${0.5*pi + 8.5/180*pi}"/>
```

Comment:
I am not sure we need a optical joint that is not the same as the base_link?
The optical joint's frame should ideally just be the same frame as the points in the pointcloud is given in, which is a fixed point in the camera - I can figure out exactly how it's specified.

Then the only frame that is essential is that optical frame, and a hand-eye transform will be used to get the pointcloud a robot's frame.
I think might be useful to also have a rough estimate of the projector coordinate system relative to the optical frame, like you have added (discussed in https://github.com/zivid/zivid-ros/pull/17/commits/624a9779cfef04e1171a1c857970b138112a0d96#r563312884).

Reply:
> The optical joint's frame should ideally just be the same frame as the points in the pointcloud is given in, which is a fixed point in the camera - I can figure out exactly how it's specified.

**joint != frame**.  
A frame is a coordinate frame. A joint is a connection between frames defining the transformation between them. 

The measurement frame (`optical_frame`) is indeed defined by the frame in which the camera outputs the captures. This may or may not coincide with another frame, but is definitely a distinct frame (even if it is just for semantics). The joint just connect the two links together.

Having a confirmation on the location of the `optical_frame` relative to the mounting hole (/`base_link`) would be very helpful indeed. Our usage (attached to a robot manipulator) does show that this location is correct or at least really close to the actual measurement frame. We often use this description "as is" without calibration for some quick captures. 

> Then the only frame that is essential is that optical frame

If looking at the camera in isolation yes, but my intent behind making this package is to actually connect it to other hardware. Then the `base_link` is essential as well, even if only by convention, expectations, and ease of use.

The `base_link` is located such that the geometry can easily be attached, it is the "starting point" of the geometry. In this case, I picked the center mounting hole as I saw this as a convenient location to which I can attach the camera for example a robot or end-effector. All description packages should start with a `base_link`.

> and a hand-eye transform will be used to get the pointcloud a robot's frame

I would say that calibration is indeed needed for real-world applications, but not part of the scope of this package. Description packages are just there to give the ideal geometry and required frames of hardware. This can be used then for simulations or for a first best guess of your real-world counterpart.

Typically calibrations will result in a new frame, for example: `calibrated_optical_frame`, that is then separately attached to the description by the user.

Reply:
**Let's keep the `base_link` link and `optical` joint. I see your point on that being useful for the simulation and a first best guess or starting point.**

> Typically calibrations will result in a new frame, for example: calibrated_optical_frame, that is then separately attached to the description by the user.

Yes, I agree, in a real-world application the hand-eye calibration will take over, to be able to know how the point cloud is related to the robot's base. And then the transformation between the `base_link` frame and the `optical_frame` is mostly useful for simulations and verifying that the robot-camera calibration is sound.

Reply:
> Having a confirmation on the location of the optical_frame relative to the mounting hole (/base_link) would be very helpful indeed.

Yes, I will get this information

Reply:
> So this will vary for each camera, even within the same model, for instance Zivid One+ M.

Would the driver have a way of retrieving that information?

There's no requirement for the `xacro:macro` to contain that link.

If the driver could publish it (as a TF frame), that would work just as well.

Reply:
> > So this will vary for each camera, even within the same model, for instance Zivid One+ M.
>
> Would the driver have a way of retrieving that information?

Would also be interested in this, especially if it is moving between usages (e.g. due to temperature differences)

Reply:
(forgive me if this has been discussed before / is generally known about Zivid devices)

Unless the pointcloud / depth images are automatically transformed by the driver to have their origins at a fixed point (so the driver / embedded firmware compensates for the offsets/variation due to temperature/other factors), not having the precise location of the optical frame significantly complicates using Zivid cameras for really precise/accurate work.

Extrinsic calibrations could likely compensate for that offset (they would incorporate it into whatever transform they determine between the camera itself and the mounting link), but IIUC from the comments by @runenordmo, that would essentially only be the extrinsic calibration for one particular 'state' of the sensor.

If the camera itself already compensates for this, a static `link` in the URDF/`xacro:macro` would seem to suffice. If not, the driver would ideally publish the transform itself -- perhaps not continuously, but at least the one associated with a particular capture. The rest of the system could then match it based on time from the `header.stamp` and the TF buffer.